synapse/storage/data_stores/main/ui_auth.py

```diff
@@ -0,0 +1,294 @@
+            description:
+                A string description of the operation that the current
+                authentication is authorising.
+
+        Returns:
+            The newly created session.
+
+        """
+        # The clientdict gets stored as JSON.
+        clientdict_json = json.dumps(clientdict)
+
+        # autogen a session ID and try to create it. We may clash, so just
+        # try a few times till one goes through, giving up eventually.
+        attempts = 0
+        while attempts < 5:
+            session_id = stringutils.random_string(24)
+
+            try:
+                await self.db.simple_insert(
+                    table="ui_auth_sessions",
+                    values={
+                        "session_id": session_id,
+                        "clientdict": clientdict_json,
+                        "uri": uri,
+                        "method": method,
+                        "description": description,
+                        "serverdict": "{}",
+                        "creation_time": self.hs.get_clock().time_msec(),
+                    },
+                    desc="create_ui_auth_session",
+                )
+                return UIAuthSessionData(
+                    session_id, clientdict, uri, method, description
+                )
+            except self.db.engine.module.IntegrityError:
+                attempts += 1
+        raise StoreError(500, "Couldn't generate a session ID.")
+
+    async def get_ui_auth_session(self, session_id: str) -> UIAuthSessionData:
+        """Retrieve a UI auth session.
+
+        Args:
+            session_id: The ID of the session.
+        Returns:
+            A dict containing the device information.
+        Raises:
+            SynapseError: if the session is not found.
+        """
+        try:
+            result = await self.db.simple_select_one(
+                table="ui_auth_sessions",
+                keyvalues={"session_id": session_id},
+                retcols=("clientdict", "uri", "method", "description"),
+                desc="get_ui_auth_session",
+            )
+        except StoreError:
+            raise SynapseError(400, "Unknown session ID: %s" % session_id)
+
+        result["clientdict"] = json.loads(result["clientdict"])
+
+        return UIAuthSessionData(session_id, **result)
+
+    async def mark_ui_auth_stage_complete(
+        self, session_id: str, stage_type: str, result: Union[str, bool, JsonDict],
+    ):
+        """
+        Mark a session stage as completed.
+
+        Args:
+            session_id: The ID of the corresponding session.
+            stage_type: The completed stage type.
+            result: The result of the stage verification.
+        """
+        await self.db.runInteraction(
+            "mark_ui_auth_stage_complete",
+            self._mark_ui_auth_stage_complete,
+            session_id,
+            stage_type,
+            result,
+        )
+
+    def _mark_ui_auth_stage_complete(
```

Comment:
convention is that we stick a `_txn` on the end of the name for things that are designed to be wrapped in `runInteraction`.
```suggestion
    def _mark_ui_auth_stage_complete_txn(
```